### PR TITLE
Add set difference method and couple of tests

### PR DESCRIPTION
**Pull request by alexeysofin**
1. Add set difference method
2. Add tests for duplicates, distinct and set difference

### CODE REVIEW

File: filter_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+func TestDifference(t *testing.T) {
+	var (
+		cases = []struct {
+			want []string
+			a    []string
+			b    []string
+		}{
+			{want: []string{"a"}, a: []string{"a", "b"}, b: []string{"b"}},
+			{want: []string{}, a: []string{"a", "b"}, b: []string{"b", "a"}},
+			{want: []string{"a"}, a: []string{"a", "a", "b"}, b: []string{"b"}},
```
Comment (sergeydobrodey) [inline filter_test.go:17]: Не уверен, что это очевидный результат для функции  Difference, если не заглядывать в детали описания. Может не будем делать Distinct?

Comment (alexeysofin) [reply filter_test.go:17]: ну вообще согласен, но тут момент что, скажем, меняет `Distinct` меняет порядок элементов, отсюда подразумевается что мы оперируем "сетами", поэтому и я подумал что тут тоже сеты нужны.

File: filter.go [hunk 2]
```diff
@@ -44,3 +44,18 @@ func Distinct[T comparable](source []T) []T {
...
+		ok   bool
+		bMap = SliceToMap(b, func(t T) T { return t })
+	)
+
+	return FilterBy(a, func(t T) bool {
+		_, ok = bMap[t]
+		// we need a elements that are not in B
+		return !ok
+	})
+}
```
Comment (sergeydobrodey) [inline filter.go:61]: Can be rewritten as:
```golang
func Difference[T comparable](a []T, b []T) []T {
  var notInB = func(item T) bool { return !InFilter(b))

  return return FilterBy(a, notInB)
}
```

Comment (alexeysofin) [reply filter.go:61]: Are you sure? `InFilter` does not even return a boolean, it's a filter function

Comment (sergeydobrodey) [reply filter.go:61]: fixed.
```golang
func Difference[T comparable](a []T, b []T) []T {
	var notInB = func(item T) bool { return !InFilter(b)(item) }

	return FilterBy(a, notInB)
}
```

However this approach is not effective - InFilter(b) is called multiple times instea

Comment (alexeysofin) [reply filter.go:61]: not sure this adds anything tgh, maybe leave as is?

File: filter.go [hunk 3]
```diff
@@ -48,14 +48,11 @@ func Distinct[T comparable](source []T) []T {
...
-	return FilterBy(a, func(t T) bool {
-		_, ok = bMap[t]
-		// we need a elements that are not in B
-		return !ok
-	})
+	var bMap = make(map[T]struct{}, len(b))
+
+	for _, v := range b {
+		bMap[v] = struct{}{}
+	}
```
Comment (sergeydobrodey) [inline filter.go:55]: unused code

Comment (alexeysofin) [reply filter.go]: sorry, a typo, fixed
